File: Java/Hybernate/example/src/main/java/example/Service/EntityService.java
package example.Service;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.List;

import example.EntityManager.EntityManager;
import example.Enums.Position;
import example.Models.Department;
import example.Models.Employee;
import example.Models.Person;

public class EntityService {

    private EntityManager<Department> departmentEManager;
    private EntityManager<Person> personEManager;
    private EntityManager<Employee> employeeEManager;

    public EntityService(
        EntityManager<Department> departmentEManager,
        EntityManager<Person> personEManager,
        EntityManager<Employee> employeeEManager
    ) {
        this.departmentEManager = departmentEManager;
        this.personEManager = personEManager;
        this.employeeEManager = employeeEManager;
    }

    public Department addNewDepartment(String name) {
        Department department = new Department(name);
        departmentEManager.addNew(department);
        return department;
    }

    public Person addNewPerson(String name, String address, int birthYear, int birthMon, int birthDay) {
        Calendar calendar = new GregorianCalendar(birthYear, birthMon-1, birthDay, 0, 0);
        Date date = calendar.getTime();
        Person person = new Person(name, address, date);
        personEManager.addNew(person);
        return person;
    }

    public Employee addNewEmployee(String personId, Long departmentId, Position position) {
        Person person = personEManager.getById(personId);
        Department department = departmentEManager.getById(departmentId);
        Employee employee = new Employee(person, department, position);
        employeeEManager.addNew(employee);
        return employee;
    }

    public Employee addNewEmployee(Department department, String name, String address, int birthYear, int birthMon, int birthDay, Position position) {
        Person person = addNewPerson(name, address, birthYear, birthMon, birthDay);
        Employee employee = addNewEmployee(person.getId(), department.getId(), position);
        return employee;
    }

    @SuppressWarnings("unchecked")
    public <T> T getById(Class<T> entityClass, Object id) throws ClassCastException {
        if (entityClass == Department.class) {
            return (T) departmentEManager.getById(id);
        }
        else if (entityClass == Person.class) {
            return (T) personEManager.getById(id);
        }
        else if (entityClass == Employee.class) {
            return (T) employeeEManager.getById(id);
        }

        throw new ClassCastException(entityClass.toString());
    }

    @SuppressWarnings("unchecked")
    public <T> List<T> getAll(Class<T> entityClass) throws ClassCastException {
        if (entityClass == Department.class) {
            return (List<T>) departmentEManager.getAll();
        }
        else if (entityClass == Person.class) {
            return (List<T>) personEManager.getAll();
        }
        else if (entityClass == Employee.class) {
            return (List<T>) employeeEManager.getAll();
        }

        throw new ClassCastException(entityClass.toString());
    }

    public <T> void update(T entity) throws ClassCastException {
        if (entity instanceof Department) {
            departmentEManager.saveChanges((Department) entity);
        }
        else if (entity instanceof Person) {
            personEManager.saveChanges((Person) entity);
        }
        else if (entity instanceof Employee) {
            employeeEManager.saveChanges((Employee) entity);
        }
        else
            throw new ClassCastException(entity.getClass().toString());
    }

    public <T> void delete(T entity) throws ClassCastException {
        if (entity instanceof Department) {
            departmentEManager.delete((Department) entity);
        }
        else if (entity instanceof Person) {
            personEManager.delete((Person) entity);
        }
        else if (entity instanceof Employee) {
            employeeEManager.delete((Employee) entity);
        }
        else
            throw new ClassCastException(entity.getClass().toString());
    }

    public <T> void deleteRange(List<T> entities) throws ClassCastException {
        for (Object entity : entities) {
            delete(entity);
        }
    }
}

File: Kotlin/Hibernate/src/main/kotlin/Services/EntityService.kt
package Services

import EntityManagers.EntityManager
import Enums.Position
import Models.Department
import Models.Employee
import Models.Person
import java.util.Calendar
import java.util.GregorianCalendar
import kotlin.ClassCastException

class EntityService(
    private val departmentEntityManager: EntityManager<Department>,
    private val personEntityManager: EntityManager<Person>,
    private val employeeEntityManager: EntityManager<Employee>
) {
    fun addNewDepartment(name: String): Department {
        val department = Department(name)
        departmentEntityManager.addNew(department)
        return department
    }

    fun addNewPerson(name: String, address: String, birthYear: Int, birthMonth: Int, birthDay: Int): Person {
        val calendar = GregorianCalendar(birthYear, birthMonth-1, birthDay)
        val date = "${calendar.get(Calendar.YEAR)}-${calendar.get(Calendar.MONTH)+1}-${calendar.get(Calendar.DAY_OF_MONTH)}"
        val person = Person(name, address, date)
        personEntityManager.addNew(person)
        return person
    }

    fun addNewEmployee(personId: String, departmentId: Long, position: Position): Employee {
        val person = personEntityManager.getById(personId)
        val department = departmentEntityManager.getById(departmentId)
        val employee = Employee(person, department, position)
        employeeEntityManager.addNew(employee)
        return employee
    }

    fun addNewEmployee(department: Department, name: String, address: String, birthYear: Int, birthMonth: Int, birthDay: Int, position: Position): Employee {
        val person = addNewPerson(name, address, birthYear, birthMonth, birthDay)
        val employee = addNewEmployee(
            person.id ?: throw NullPointerException("person.id"),
            department.id ?: throw NullPointerException("department.id"),
            position
        )

        return employee
    }

    fun <T> getById(entityClass: Class<T>, id: Any): T = when (entityClass) {
        Department::class.java -> departmentEntityManager.getById(id) as T
        Person::class.java -> personEntityManager.getById(id) as T
        Employee::class.java -> employeeEntityManager.getById(id) as T
        else -> throw ClassCastException(entityClass.toString())
    }

    fun <T> getAll(entityClass: Class<T>): List<T> = when (entityClass) {
        Department::class.java -> departmentEntityManager.getAll() as List<T>
        Person::class.java -> personEntityManager.getAll() as List<T>
        Employee::class.java -> employeeEntityManager.getAll() as List<T>
        else -> throw ClassCastException(entityClass.toString())
    }

    fun <T> update(entity: T) = when (entity) {
        is Department -> departmentEntityManager.saveChanges(entity)
        is Person -> personEntityManager.saveChanges(entity)
        is Employee -> employeeEntityManager.saveChanges(entity)
        else -> throw ClassCastException(entity!!::class.java.toString())
    }

    fun <T> delete(entity: T) = when (entity) {
        is Department -> departmentEntityManager.delete(entity)
        is Person -> personEntityManager.delete(entity)
        is Employee -> employeeEntityManager.delete(entity)
        else -> throw ClassCastException(entity!!::class.java.toString())
    }

    fun <T> deleteRange(entities: List<T>) = entities.forEach { delete(it) }
}
